### PR TITLE
Hw2

### PR DESCRIPTION
**Pull request by Julia-92**

### CODE REVIEW

File: pages/home.js
Comment (adudyak) [inline pages/home.js:1]: move methods, related to product - to productPage object, checkout - checkoutPage, cart - cart page

File: pages/home.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  },
+
+  clickConfirmOrderButton() {
+    I.click(this.confirmOrderButton);
+  },
+  
+
+  //clickRemoveItems() {
+  //  I.click(this.removeItems);
+  //},
```
Comment (adudyak) [inline pages/home.js]: remove commented code

File: tests/buyProduct_test.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    homePage.clickContinueButton();
+    homePage.clickAgree();
+    homePage.clickContinueButton();
+    //homePage.clickContinueButton();
+    //homePage.clickAgree();
+
+    //homePage.clickContinueButton();
+    homePage.clickConfirmOrderButton();
+   
+    pause();
```
Comment (adudyak) [inline tests/buyProduct_test.js]: remove pause()

File: tests/buyProduct_test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+
+Feature('purchase');
+
+Scenario.only('buy product',  ({ I, homePage }) => {
+   
+    I.login(USER);
+    I.amOnPage('http://opencart.qatestlab.net/index.php?route=product/product&product_id=74');
+    //homePage.clickDropdownCartIcon();
```
Comment (adudyak) [inline tests/buyProduct_test.js]: remove commented code

File: tests/buyProduct_test.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    //homePage.chooseCountry();
+    //homePage.clickRegionToggle();
+    //homePage.chooseRegion();
+    homePage.clickContinueButton();
+    homePage.clickContinueButton();
+    homePage.clickContinueButton();
+    //I.waitForVisible({xpath: '//label[@for="agree1"]'});
+    homePage.clickContinueButton();
+    homePage.clickAgree();
+    homePage.clickContinueButton();
```
Comment (adudyak) [inline tests/buyProduct_test.js]: you can hide all these Comtinue clicks under single method, e.g. completeSteps1-4

File: tests/buyProduct_test.js
Comment (adudyak) [inline tests/buyProduct_test.js:1]: compare product price at product page and checkout page

**Review by adudyak — CHANGES_REQUESTED** [review]

File: tests/buyProduct_test.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    const flatShippingRate = await I.grabTextFrom({xpath: '(//tfoot/tr/td[@class="text-right"])[4]'});
+    const flatShippingRateNum = +flatShippingRate.slice(1);
+    console.log(flatShippingRateNum);
+    const totalPrice = await I.grabTextFrom({xpath: '(//tfoot/tr/td[@class="text-right"])[6]'});
+    const totalPriceNum = +totalPrice.slice(1);
+    console.log(totalPriceNum);
+
+    I.assertEqual(itemPriceNum+flatShippingRateNum, totalPriceNum, 'prices are not in match');
+    homePage.clickConfirmOrderButton();
+    I.see('Your order has been placed!');
```
Comment (adudyak) [inline tests/buyProduct_test.js]: move verification to checkout page

**Review by adudyak — CHANGES_REQUESTED** [review]

File: tests/buyProduct_test.js (hunk 2 above)
Comment (Julia-92) [reply tests/buyProduct_test.js]: > remove pause()

++

File: pages/home.js (hunk 1 above)
Comment (Julia-92) [reply pages/home.js]: > remove commented code

++

File: pages/home.js
Comment (Julia-92) [reply pages/home.js:1]: > cart page

did not exist a cart page file, in the homePage placed

File: tests/buyProduct_test.js (hunk 3 above)
Comment (Julia-92) [reply tests/buyProduct_test.js]: > remove commented code

++

File: tests/buyProduct_test.js (hunk 4 above)
Comment (Julia-92) [reply tests/buyProduct_test.js]: added loop

Comment (Julia-92) [reply tests/buyProduct_test.js]: > you can hide all these Comtinue clicks under single method, e.g. completeSteps1-4

added loop

File: tests/buyProduct_test.js
Comment (Julia-92) [reply tests/buyProduct_test.js:1]: > compare product price at product page and checkout page

++

File: tests/buyProduct_test.js (hunk 5 above)
Comment (Julia-92) [reply tests/buyProduct_test.js]: > move verification to checkout page

++

File: pages/checkout.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  continueButton: {xpath: '(//input[@value="Continue"])[last()]'},
+  agree: {xpath: '//input[@name="agree"]'},
+  confirmOrderButton: {xpath: '//input[@value="Confirm Order"]'},
+
+ 
+  itemPriceText: {xpath: '(//tbody/tr/td)[last()]'},
+  flatShippingRateText: {xpath: '(//tfoot/tr/td[@class="text-right"])[4]'},
+  totalPriceText: {xpath: '(//tfoot/tr/td[@class="text-right"])[6]'},
+
+  fillCheckoutForm2(checkout) {
```
Comment (adudyak) [inline pages/checkout.js:24]: name it more specifically, like `fillBillingForm()` or `...shipping...`

File: pages/checkout.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    I.click(this.agree);
+  },
+
+  clickConfirmOrderButton() {
+    I.click(this.confirmOrderButton);
+  },
+
+  verifyOrderPageName() {
+    I.see('Your order has been placed!');
+  },
```
Comment (adudyak) [inline pages/checkout.js:62]: group these 1 row functions under one/two functions.

File: pages/checkout.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+  async grabItemPrice() {
```
Comment (adudyak) [inline pages/checkout.js:64]: you don't need to grab item price at checkout page, only at product page

File: pages/checkout.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    const ItemPrice = await I.grabTextFrom(this.itemPriceText);
+    const numItemPrice = +ItemPrice.slice(1);
+    return numItemPrice;
+  },
+  
+  async grabFlatShippingRate() {
+    const FlatShippingRate = await I.grabTextFrom(this.flatShippingRateText);
```
Comment (adudyak) [inline pages/checkout.js:71]: use lowercase variable

File: pages/checkout.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    const numFlatShippingRate = +FlatShippingRate.slice(1);
+    return numFlatShippingRate;
+  },
+
+  async grabTotalPrice() {
+    const TotalPrice = await I.grabTextFrom(this.totalPriceText);
```
Comment (adudyak) [inline pages/checkout.js:77]: use lowercase variable

File: pages/product.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+const { I } = inject();
+
+module.exports = {
+
+  selectField: {xpath: '//div/a[.="--- Please Select ---"]'},
+  colorField: {xpath: '(//ul[@class="sbOptions"]/li/a)[2]'},
+  addToCartButton: {xpath: '//button[@id="button-cart"]'},
+  priceInProductPageText: {xpath: '//div[@class="price"]/span'},
+
+  clickSelectField() {
```
Comment (adudyak) [inline pages/product.js:10]: add `I.click(this.colorField);` to this method and name it `selectColor()`

File: pages/product.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    I.click(this.selectField);
+  },
+
+  clickColor() {
```
Comment (adudyak) [inline pages/product.js:14]: remove this one

File: pages/product.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    I.click(this.colorField);
+  },
+
+  clickAddToCartButton() {
+    I.click(this.addToCartButton);
+  },
+
+  async grabPriceInProductPage() {
+    const PriceInProductPage = await I.grabTextFrom(this.priceInProductPageText);
```
Comment (adudyak) [inline pages/product.js:23]: lowercase variable

File: tests/buyProduct_test.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    console.log(itemPrice);
+    const flatShippingRate = await checkoutPage.grabFlatShippingRate();
+    console.log(flatShippingRate);
+    const totalPrice = await checkoutPage.grabTotalPrice();
+    console.log(totalPrice);
+    I.assertEqual(itemPrice+flatShippingRate, totalPrice, 'prices are not in match');
+    checkoutPage.clickConfirmOrderButton();
+    checkoutPage.verifyOrderPageName();
+
+    I.amOnPage('http://opencart.qatestlab.net/index.php?route=product/product&product_id=74');
```
Comment (adudyak) [inline tests/buyProduct_test.js:50]: remove duplicated code 50-54

File: tests/buyProduct_test.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    const priceInProductPage = await productPage.grabPriceInProductPage();
+    const colorPriceInProductPage = await productPage.grabColorPrice();
```
Comment (adudyak) [inline tests/buyProduct_test.js:52]: extract price and colorPrice before proceeding to checkout

File: tests/buyProduct_test.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    I.amOnPage('http://opencart.qatestlab.net/index.php?route=product/product&product_id=74');
+    productPage.clickSelectField();
+    productPage.clickColor();
+    productPage.clickAddToCartButton();
+    homePage.clickDropdownCartIcon();
+    homePage.clickCheckout();
+    checkoutPage.fillCheckoutForm2(USER);
+    checkoutPage.clickCountryToggle();
+
+    for (let i = 0; i < 4; i++) {
```
Comment (adudyak) [inline tests/buyProduct_test.js:33]: don't hardcode cycle to 4, use `while` cycle with condition: button is visible or similar

File: tests/buyProduct_test.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    postcode: "61072",
+}
+
+
+Feature('purchase');
+
+Scenario.only('buy product',  async ({ I, homePage, checkoutPage, productPage }) => {
+   
+    I.login(USER);
+    homePage.clickDropdownCartIcon();
```
Comment (adudyak) [inline tests/buyProduct_test.js:18]: move logic to clear cart to home page in `clearCart()` method.

File: pages/checkout.js (hunk 6 above)
Comment (Julia-92) [reply pages/checkout.js:24]: > name it more specifically, like `fillBillingForm()` or `...shipping...`

++

File: pages/checkout.js (hunk 7 above)
Comment (Julia-92) [reply pages/checkout.js:62]: > group these 1 row functions under one/two functions.

I created verifyPage function in the homePage and changed call at register_test also.

File: pages/checkout.js (hunk 8 above)
Comment (Julia-92) [reply pages/checkout.js:64]: > you don't need to grab item price at checkout page, only at product page

I left only one assertEqual

File: pages/checkout.js (hunk 9 above)
Comment (Julia-92) [reply pages/checkout.js:71]: > use lowercase variable

++

File: pages/checkout.js (hunk 10 above)
Comment (Julia-92) [reply pages/checkout.js:77]: > use lowercase variable

++

File: pages/product.js (hunk 13 above)
Comment (Julia-92) [reply pages/product.js:23]: > lowercase variable

++

File: tests/buyProduct_test.js (hunk 14 above)
Comment (Julia-92) [reply tests/buyProduct_test.js:50]: > remove duplicated code 50-54

I don't find it, maybe already deleted

File: tests/buyProduct_test.js (hunk 15 above)
Comment (Julia-92) [reply tests/buyProduct_test.js:52]: > 

++

File: tests/buyProduct_test.js (hunk 16 above)
Comment (Julia-92) [reply tests/buyProduct_test.js:33]: > don't hardcode cycle to 4, use `while` cycle with condition: button is visible or similar

Done. But the Agree button is not clicked when the 4th ContinueButton was

**Pull request merged**